### PR TITLE
updated workflow to enable caching of ESMF

### LINKED ISSUE #260: Modify existing workflow to use caches for ESMF, NCEPLIBS, etc.
Currently we are rebuilding the same versions of ESMF, Jasper, and NCEPLIBS.

Seems like we could cache these, since they are the same every time. This would make the CI faster.

### PR DESCRIPTION
**Pull request by edwardhartnett**
Fifth time's the charm!

Part of #260 

It took me many tries to get the right. Turns out that GitHub actions sometimes just freezes when your workflow file is wrong. ;-)

In this PR I change the CI to get ESMF and other codes with wget, instead of checking them out, and then I use GitHub caching, so that these codes are only build the first time. For subsequent tests, GitHub pulls them from the cache.

### CODE REVIEW

**Comment by edwardhartnett** [comment]
OK, this is ready to merge and is an improvement. @kgerheiser you may want to take a look to see if you think I messed anything up here. Basically, I moved where ESMF was installed, cached it, and now use a wget to get a tarball instead of checking out a tagged version of the ESMF code.

The end result is that ESMF will no longer be built each time the test is run.

After this PR is merged I will take another stab and doing this for the jasper and nceplibs in this CI as well. But after all this dorking around, I want to at least get the ESMF part into develop branch! ;-)

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -77,7 +84,8 @@ jobs:
 
     - name: build-ufs-utils
       run: |
-        export ESMFMKFILE=~/lib/esmf.mk
+        export ESMFMKFILE=~/esmf/lib/esmf.mk
+        export ESMF_DIR=~/esmf-ESMF_8_0_1
```
Comment (kgerheiser) [inline .github/workflows/build_and_test.yml:88]: Why do you need to set ESMF_DIR here?

Comment (edwardhartnett) [reply .github/workflows/build_and_test.yml:88]: Maybe I don;t. Let me test it...

Comment (edwardhartnett) [reply .github/workflows/build_and_test.yml]: I don't!

**Review by kgerheiser — APPROVED** [review]

**Comment by GeorgeGayno-NOAA** [comment]
I see you are now saving 10 minutes of compile time.  Nice.

**Review by GeorgeGayno-NOAA — APPROVED** [review]

**Pull request merged**